iotfunctions/bif.py

```diff
@@ -109,6 +109,176 @@ def execute(self, x):
+        return y
+
+class StateTimePreparation(BaseTransformer):
+    '''
+    Together with AggregateTimeInState StateTimePreparation
+    calculates the amount of time a selected metric has been in a
+    particular state.
+    StateTimePreparation outputs an encoded pair of a state change
+    variable (-1 for leaving the state, 0 for no change,
+     1 for entering the state) together with a unix epoch
+    timestamp.
+    The condition for the state change is given as binary operator
+    together with the second argument, for example
+    ">= 37"  ( for fever) or "=='running'" (for process states)
+    '''
+    def __init__(self, source=None, state_name=None, name=None):
+        super().__init__()
+        logger.info('StateTimePrep _init')
+
+        self.source = source
+        self.state_name = state_name
+        self.name = name
+        print(dir(self))
+
+    @classmethod
+    def build_ui(cls):
+        inputs = []
+        inputs.append(UISingleItem(name='source', datatype=float,
+                                  description='Data item to compute the state change array from'))
+        inputs.append(UISingle(name='state_name', datatype=str,  description='Condition for the state change array computation
+
+        outputs = []
+        outputs.append(
+            UIFunctionOutSingle(name='name', datatype=str, description='State change array output'))
+
+        return (inputs, outputs)
+
+    def _calc(self, df):
+        logger.info('Execute StateTimePrep per entity')
+
+        index_names = df.index.names
+        ts_name = df.index.names[1]  # TODO: deal with non-standard dataframes (no timestamp)
+
+        print('Source ', self.source, 'state_name ', self.state_name, 'Name ', self.name)
+        #df[self.name] = (df[self.source] == self.state_name).astype(int).diff().fillna(1).astype(int)
+        df_copy = df.reset_index()
+
+        # pair of +- seconds and regular timestamp
+        v1 = eval("df_copy[self.source] " + self.state_name).astype(int).diff().values.astype(int)
+        #v1 = (df_copy[self.source] > 50).astype(int).diff().values.astype(int)
+        # first element is NaN
+        if v1.size > 0:
+            v1[0] = 0
+            try:
+                nonzero = np.min(np.nonzero(v1 != 0))
+                if v1[nonzero] > 0:
+                    v1[0] = -1
+                else:
+                    v1[0] = 1
+            except Exception:
+                # no non zero element
+                pass
+
+        df_copy['__intermediate1__'] = v1
+        np.savetxt('/tmp/test', df_copy['__intermediate1__'].values)
+        df_copy['__intermediate2__'] = (df_copy[ts_name].astype(int)// 1000000000)
+        df_copy[self.name] = df_copy['__intermediate1__'].map(str) + ',' + df_copy['__intermediate2__'].map(str)
+
+        df_copy.drop(columns=['__intermediate1__','__intermediate2__'], inplace=True)
+        df_copy.to_csv('/tmp/testc')
```

Comment:
Are `tmp/test` and `tmp/testc` necessary to create and keep around? Are these files actually used anywhere or are they just leftover from debugging?

Reply:
I left them in for debugging - /tmp is available on a docker container and is recreated with every pipeline run.
but it's safer to delete them (same for the np.saveext in the same code block)

Reply:
Pushed the changes you suggested